urlb44e9843ba_QuasimodoNZ_SWEN222_Game_Project_tgz-pJ8-Game_World_GameWorldLevelBuilderJ8/src/Graphics_Renderer/MapEditor.java

```diff
@@ -0,0 +1,389 @@
+		// Creates the dialog and adds the panel and its grid to go in it.
+		final JDialog dialog = new JDialog(frame, "Layer Editor");
+		dialog.setLocation(200, 350);
+		dialog.setModal(true);
+
+		JPanel panel = new JPanel(new GridBagLayout());
+		GridBagConstraints grid = new GridBagConstraints();
+		grid.insets = new Insets(0, 0, 0, 0);
+
+		textFields = new ArrayList<List<JTextField>>();
+		for (int y = 0; y < selectedItem.getIDs().length; y++) {
+
+			List<JTextField> row = new ArrayList<JTextField>();
+			for (int x = 0; x < selectedItem.getIDs()[0].length; x++) {
+				final JTextField ID = new JTextField(2);
+				ID.setText(String.format("%d", selectedItem.getIDs()[y][x]));
+				row.add(ID);
+				ID.addFocusListener(new java.awt.event.FocusAdapter() {
+					@Override
+					public void focusGained(java.awt.event.FocusEvent evt) {
+						SwingUtilities.invokeLater(new Runnable() {
+
+							@Override
+							public void run() {
+								ID.selectAll();
+							}
+						});
+					}
+				});
+			}
+			textFields.add(row);
+		}
+
+		JButton submit = new JButton("Submit");
+		submit.addActionListener(new ActionListener() {
+			@Override
+			public void actionPerformed(ActionEvent e) {
+
+				for (int i = 0; i < selectedItem.getIDs().length; i++)
+					for (int j = 0; j < ((ImageLayer) (imageLayers
+							.getSelectedItem())).getIDs()[0].length; j++) {
+						selectedItem.getIDs()[i][j] = Integer
+								.parseInt(textFields.get(i).get(j).getText());
+					}
+				// Needs to dispose of the dialog upon submitting
+				dialog.dispose();
+			}
+		});
+		grid.gridwidth = 1;
+		grid.gridheight = 1;
+		for (int y = 0; y < selectedItem.getIDs().length; y++)
+			for (int x = 0; x < selectedItem.getIDs()[y].length; x++) {
+				grid.gridx = x;
+				grid.gridy = y;
+				panel.add(textFields.get(y).get(x), grid);
+			}
+
+		// Adds the submit button
+		grid.gridwidth = textFields.get(0).size() / 3;
+		grid.gridx = textFields.get(0).size() / 3;
+		grid.gridy = textFields.size() + 1;
+		panel.add(submit, grid);
+
+		dialog.add(panel);
+		dialog.pack();
+		dialog.setVisible(true);
+		return dialog;
+	}
+
+	private void redraw(Graphics g) {
+		g.translate((int) dx, (int) dy);
+
+		// Draws each layer on the image
+		for (int i = 0; i < imageLayers.getItemCount(); i++) {
+			g.drawImage(imageLayers.getItemAt(i).toImage(), 0, 0, null);
+		}
+		System.out.printf("Graphics translated - dx: %f, dy: %f\n\n", dx, dy);
+
+	}
+
+	private void createOutputFile() {
+		// Prepare browse dialog
+		JFileChooser fc = new JFileChooser();
+		fc.setFileHidingEnabled(false);
+		fc.setAcceptAllFileFilterUsed(false);
+		fc.setApproveButtonText("Open");
+
+		// Show dialog
+		int rt = fc.showOpenDialog(frame); // someframe is JFrame
+		if (rt == JFileChooser.APPROVE_OPTION) {
+			File file = fc.getSelectedFile(); // Do anything u want with this
+												// file
+			BufferedWriter writer;
+			try {
+				writer = new BufferedWriter(new FileWriter(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *writer*. The resource is referenced at lines 351, 354, 355, 356, 358, 360, 363, 365, 367, 369, 370, 371, 372, 373. The resource is closed at line 376. There are other execution paths that do not contain closure statements, for example, when BufferedWriter.newLine throws an exception. Ensure *writer* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).